pkg/plugins/pfs/pfs.go

```diff
@@ -55,250 +81,168 @@ func init() {
 	}
+	if rt == nil {
+		rt = cuectx.GrafanaThemaRuntime()
+	}
+
 	lin, err := plugindef.Lineage(rt)
 	if err != nil {
 		panic(fmt.Sprintf("plugindef lineage is invalid or broken, needs dev attention: %s", err))
 	}
-	mux := vmux.NewValueMux(lin.TypedSchema(), vmux.NewJSONCodec("plugin.json"))
 	ctx := rt.Context()
 
-	b, err := fs.ReadFile(f, "plugin.json")
+	b, err := fs.ReadFile(fsys, "plugin.json")
 	if err != nil {
 		if errors.Is(err, fs.ErrNotExist) {
-			return nil, ErrNoRootFile
+			return ParsedPlugin{}, ErrNoRootFile
 		}
-		return nil, fmt.Errorf("error reading plugin.json: %w", err)
+		return ParsedPlugin{}, fmt.Errorf("error reading plugin.json: %w", err)
 	}
 
-	tree := &Tree{
-		raw: f,
-		rootinfo: PluginInfo{
-			slotimpls: make(map[string]thema.Lineage),
-		},
+	pp := ParsedPlugin{
+		ComposableKinds: make(map[string]kindsys.Composable),
+		// CustomKinds:     make(map[string]kindsys.Custom),
 	}
-	r := &tree.rootinfo
 
 	// Pass the raw bytes into the muxer, get the populated PluginDef type out that we want.
 	// TODO stop ignoring second return. (for now, lacunas are a WIP and can't occur until there's >1 schema in the plugindef lineage)
-	pmeta, _, err := mux(b)
+	codec := vmux.NewJSONCodec("plugin.json")
+	pinst, _, err := vmux.NewTypedMux(lin.TypedSchema(), codec)(b)
 	if err != nil {
 		// TODO more nuanced error handling by class of Thema failure
-		return nil, ewrap(err, ErrInvalidRootFile)
+		return ParsedPlugin{}, ewrap(err, ErrInvalidRootFile)
+	}
+	pp.Properties = *(pinst.ValueP())
+
+	if cuefiles, err := fs.Glob(fsys, "*.cue"); err != nil {
+		return ParsedPlugin{}, fmt.Errorf("error globbing for cue files in fsys: %w", err)
+	} else if len(cuefiles) == 0 {
+		return pp, nil
 	}
-	r.meta = *pmeta
-
-	if modbyt, err := fs.ReadFile(f, "models.cue"); err == nil {
-		// TODO introduce layered CUE dependency-injecting loader
-		//
-		// Until CUE has proper dependency management (and possibly even after), loading
-		// CUE files with non-stdlib imports requires injecting the imported packages
-		// into cue.mod/pkg/<import path>, unless the imports are within the same CUE
-		// module. Thema introduced a system for this for its dependers, which we use
-		// here, but we'll need to layer the same on top for importable Grafana packages.
-		// Needing to do this twice strongly suggests it needs a generic, standalone
-		// library.
-
-		mfs := merged_fs.NewMergedFS(f, grafana.CueSchemaFS)
-
-		// Note that this actually will load any .cue files in the fs.FS root dir in the plugindef.PkgName.
-		// That's...maybe good? But not what it says on the tin
-		bi, err := load.InstanceWithThema(mfs, "", load.Package(plugindef.PkgName))
-		if err != nil {
-			return nil, fmt.Errorf("loading models.cue failed: %w", err)
-		}
 
-		pf, _ := parser.ParseFile("models.cue", modbyt, parser.ParseComments)
+	gpv := loadGP(rt.Context())
+	pinst.Underlying()
 
-		for _, im := range pf.Imports {
+	fsys, err = ensureCueMod(fsys, pp.Properties)
```

Comment:
We're now ensuring/injecting a `cue.mod` directory into all plugins, including core ones. That means we're effectively treating core plugins as being logically outside the grafana CUE module.

i haven't seen any problems with this...yet? and it's modeling the world we want, so 🎉